### PR TITLE
Core loader

### PR DESCRIPTION
**Pull request by wpears**
Covers all the use cases of the retriever; handles all supported file types.

Not Dockerized yet to spare the reviewer of this pull request ^^;

Enjoy :boom:

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+  var scratchSpace = 'retriever-scratch' + Math.random()*1e17;
+  fs.mkdirSync(scratchSpace);
+
+
+  function wrappedCb(err, count){
+    var args = arguments;
+    var self = this;
+
+    fs.remove(scratchSpace, function(e){
```
Comment (hkeeler) [inline index.js:26]: Are you sure you want to remove the scratch space whenever an error occurs?  Any chance there are half-processed files in there you'd want to inspect before they get ditched?

File: index.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      if(err||e){
+       if(callback) return callback(err||e);
```
Comment (hkeeler) [inline index.js:28]: Indentation is off here.

Comment (wpears) [inline index.js:28]: This has been fixed by #9

File: index.js (hunk 1 above)
Comment (wpears) [inline index.js:26]: Yep, if there is an unexpected error that crashes the processes, the scratch space doesn't get cleaned up (I could have hooked into process.on('uncaughtException') or used node domains if I wanted to ensure cleanup, but for the very reason you bring up, intermediate files can be useful in the case of unexpected errors.

File: data.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
+[
```
Comment (hkeeler) [inline data.json:1]: What's the plan/workflow for keeping this file up-to-date?  Every time a new file is available, we have to update `hash`?  In some cases it looks like we'll have to update `file` too, like `AddressNC_2014.gdb`.  That's unfortunate.

Comment (wpears) [inline data.json:1]: It is unfortunate and I've already felt a taste of it. However, it saves us from weird loading problems on state-side schema changes (which has actually already happened). Essentially the workflow is to download the file, convert it to geojson and check its fields against those that the loader expects. If there is no difference, we can assume that there were just edits/additions and update the hash. If there _are_ changes to the schema, the transformer in the loader (and associated tests) need to be updated).

Currently I've been doing this by hand, but I will automate it as soon as I feel confident that I can account for edge cases.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    }else{
+      if(csvReg.test(record.file)){
+        var csv = path.join(scratchSpace, record.file);
+        var csvStream = fs.createWriteStream(csv);
+
+        csvStream.on('finish', function(err){
+          if(err) return recordCallback(err);
+
+          csvToVrt(csv, record.sourceSrs, function(err, vrt){
+            if(err) return recordCallback(err);
```
Comment (hkeeler) [inline index.js:116]: What is the expected behavior when an error occurs in the middle of processing the set of files?  Does the the entire process stop, or does it continue on to the next file?  And if it does continue on, can the `scratchSpace` get deleted out from under the next file?

Sorry, its late, and I'm having trouble following all the callback/error handling here.

Comment (wpears) [inline index.js:116]: As it stands, the `wrappedCb` will throw an error when it is called, ending the whole process. (`recordCallback`, if called with an error, passes this error on to `wrappedCb`). This actually doesn't need to happen, as long as it is clearly reported what failed and why. re: scratchSpace, the file creation code will make intermediate dirs if they don't exist, so this shouldn't be a problem. That said, there could be race conditions where a file is created then deleted by another failing request. So another reason to allow failure to be isolated and let the process continue. Cleanup would then be handled after everything has either completed successfully or failed.

I'll add this change in another PR.

Comment (hkeeler) [inline index.js:116]: Cool.  Yeah, it was the race condition scenario that seemed possible if the whole job _didn't_ stop, but since it does, less of an issue.  Of course, as we add more file, we won't want any given failure to bring the whole thing to a halt.

Sounds like you have a good plan for handling this in the future.  Stand by for mergification...
